astropy/utils/data.py

```diff
@@ -275,26 +275,13 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
     else:
         fileobj = name_or_obj
 
-    # Check if the file object supports random access, and if not,
-    # then wrap it in a BytesIO buffer.  It would be nicer to use a
-    # BufferedReader to avoid reading loading the whole file first,
-    # but that is not compatible with streams or urllib2.urlopen
-    # objects on Python 2.x.
-    if not hasattr(fileobj, 'seek'):
-        try:
-            # py.path.LocalPath objects have .read() method but it uses
-            # text mode, which won't work. .read_binary() does, and
-            # surely other ducks would return binary contents when
-            # called like this.
-            # py.path.LocalPath is what comes from the tmpdir fixture
-            # in pytest.
-            fileobj = io.BytesIO(fileobj.read_binary())
-        except AttributeError:
-            fileobj = io.BytesIO(fileobj.read())
-
     # Now read enough bytes to look at signature
-    signature = fileobj.read(4)
-    fileobj.seek(0)
+    if fileobj.seekable():
+        signature = fileobj.read(4)
+        fileobj.seek(0)
+    else:
+        fileobj = io.BufferedReader(fileobj)
+        signature = fileobj.peek(4)
```

Comment:
But if a file-like object has no `.seek` method, it will also most probably not have a `.seekable` method and this will crash ?

Reply:
If an object does not have `seek()` and `seekable()` methods then it is not an instance of [io.IOBase](https://docs.python.org/3.8/library/io.html#io.IOBase).